text/0000-yarn-berry-support.md

```diff
@@ -0,0 +1,65 @@
+
+With the addition of Yarn 3 pulling from a different package (though still using the npm registry), this could get more complicated. My current suggestion is that we keep the hooks format the same even for Yarn Berry releases. This will mean that teams using hooks will have to make sure the correct package is available at the appropriate URLs themselves. Then, in a future change where we update the hooks format, we can migrate all of Yarn to use the npm format and normalize the behavior.
+
+## Plug'n'Play
+
+Beyond installation, the other blocker for fully supporting Yarn Berry is that our global package delegation does not work with Yarn Plug'n'Play. The root issue is that we expect any project-local binaries to be inside of `node_modules/.bin`, which isn't the case for Plug'n'Play installations.
+
+To work around that, I believe we can detect the presence of the Plug'n'Play include file (named `.pnp.js` or `.pnp.cjs` depending on the Yarn version) and instead of calling the binary ourselves, delegate to it using `yarn run <binary>`. This will make launching a tool slightly slower in those scenarios, however there is no way (without tightly coupling to internal Yarn implementation details) to resolve the path to a binary in Plug'n'Play without invoking `yarn`.
+
+# Critique
+[critique]: #critique
+
+## Broader Version Support
+
+One alternative approach would be for us to attempt to support as many versions of Yarn Berry as we can, including in the 2.x major version. This would give us the greatest coverage of possible use-cases, however given the churn in distribution and API, it will be very labor-intensive to implement. It also would likely still not support every possible version of Yarn, still leaving some gaps in our coverage. Since Yarn 2.x is not recommended by Yarn themselves either at this point, the additional work to support various 2.x releases feels like it would have a low impact while significantly increasing the complexity of the change.
+
+## Hooks Format
+
+The current suggestion for the hooks is the easiest from an implementation standpoint, however it offloads that work onto users that are taking advantage of hooks. However, our Hooks behavior is already not ideal and adding more complexity to it feels like it would make things even more difficult to understand. Separate from this issue, we should investigate an improved Hooks V2, with a new file format and better ergonomics, taking into account everything we learned from the original hooks implementation. At that time, we can define a more appropriate API for the Yarn hooks. That said, I don't believe our support for Yarn Berry should be blocked on redesigning Hooks, which is likely a much bigger issue overall.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- Should we consider Corepack support as part of support for Yarn Berry? ([Corepack](https://nodejs.org/api/corepack.html) is Yarn's currently-recommended way of installing / managing Yarn versions)
```

Comment:
Would this be doable as a followup to support for Yarn Berry?

Reply:
Yeah, this can definitely be separate, especially if we have documentation on the "Volta way" of making it available.

Reply:
Yeah, I think we need to tackle Corepack separately, as it's a very different bundle of concerns and is sort of notionally a competitor to *part* of what we do in Volta.